flaky test: `TestFollowSource_HeadsDivergeThenConverge`

**Job:** `memory-all-opn-op-geth`

**CI run:** https://app.circleci.com/pipelines/github/ethereum-optimism/optimism/121197/workflows/129b5275-5553-4898-96f6-8d86c6cc9afc/jobs/4696329/tests

**Branch:** develop

**Date:** 2026-03-30

## Error

Panic during test teardown — `pebble: closed` race condition in SafeDB:

```
panic: pebble: closed

goroutine 38932 [running]:
github.com/cockroachdb/pebble.(*DB).newIter(...)
	/home/circleci/go/pkg/mod/github.com/cockroachdb/pebble@v1.1.5/db.go:1011 +0x585
github.com/ethereum-optimism/optimism/op-node/node/safedb.(*SafeDB).SafeHeadAtL1(...)
	/home/circleci/project/op-node/node/safedb/safedb.go:170 +0x1eb
github.com/ethereum-optimism/optimism/op-supernode/supernode/chain_container/virtual_node.(*simpleVirtualNode).L1AtSafeHead(...)
	/home/circleci/project/op-supernode/supernode/chain_container/virtual_node/virtual_node.go:240 +0x184
github.com/ethereum-optimism/optimism/op-supernode/supernode/chain_container.(*simpleChainContainer).safeDBAtL2(...)
	/home/circleci/project/op-supernode/supernode/chain_container/chain_container.go:403 +0x269
github.com/ethereum-optimism/optimism/op-supernode/supernode/chain_container.(*simpleChainContainer).OptimisticAt(...)
	/home/circleci/project/op-supernode/supernode/chain_container/chain_container.go:450 +0x26b
github.com/ethereum-optimism/optimism/op-supernode/supernode/activity/interop.(*Interop).l1Inclusion(...)
	/home/circleci/project/op-supernode/supernode/activity/interop/algo.go:41 +0x116
```

## Analysis

The interop activity goroutine calls `SafeHeadAtL1` on a pebble DB that has already been closed during shutdown. This is a use-after-close race — the supernode's interop activity is still running when the SafeDB is closed during teardown.

The fix likely needs to ensure the interop activity is fully stopped before closing the SafeDB, or add a closed-state check in `SafeHeadAtL1`.

--- a/op-node/node/safedb/safedb.go
+++ b/op-node/node/safedb/safedb.go
@@ -17,6 +17,7 @@ import (
 var (
 	ErrNotFound     = errors.New("not found")
 	ErrInvalidEntry = errors.New("invalid db entry")
+	ErrClosed       = errors.New("safe db closed")
 )
 
 const (
@@ -100,6 +101,9 @@ func (d *SafeDB) Enabled() bool {
 func (d *SafeDB) SafeHeadUpdated(safeHead eth.L2BlockRef, l1Head eth.BlockID) error {
 	d.m.Lock()
 	defer d.m.Unlock()
+	if d.closed {
+		return ErrClosed
+	}
 	d.log.Info("Record local safe head", "l2", safeHead.ID(), "l1", l1Head)
 	batch := d.db.NewBatch()
 	defer batch.Close()
@@ -115,6 +119,9 @@ func (d *SafeDB) SafeHeadUpdated(safeHead eth.L2BlockRef, l1Head eth.BlockID) er
 func (d *SafeDB) SafeHeadReset(safeHead eth.L2BlockRef) error {
 	d.m.Lock()
 	defer d.m.Unlock()
+	if d.closed {
+		return ErrClosed
+	}
 	d.log.Info("Resetting safe head db", "l2", safeHead.ID())
 	iter, err := d.db.NewIter(safeByL1BlockNumKey.IterRange())
 	if err != nil {
@@ -167,6 +174,10 @@ func (d *SafeDB) SafeHeadReset(safeHead eth.L2BlockRef) error {
 func (d *SafeDB) SafeHeadAtL1(ctx context.Context, l1BlockNum uint64) (l1Block eth.BlockID, safeHead eth.BlockID, err error) {
 	d.m.RLock()
 	defer d.m.RUnlock()
+	if d.closed {
+		err = ErrClosed
+		return
+	}
 	iter, err := d.db.NewIterWithContext(ctx, safeByL1BlockNumKey.IterRange())
 	if err != nil {
 		return

